### PR TITLE
UV-267 - Update http messaging to mproxy

### LINKED ISSUE #1942: `Make` cleandocker is missing profiles
**BUG REPORT**

1. What were you trying to achieve?
Stop and remove docker containers

3. What are the expected results?
stopping all containers and removing them

5. What are the received results?
```shell
sudo make cleandocker           2 ✘ 
# Stops containers and removes containers, networks, volumes, and images created by up
docker-compose -f docker/docker-compose.yml down --rmi all -v --remove-orphans
service nats is required by coap-adapter but is disabled. Can be enabled by profiles [nats_nats nats_rabbitmq]
make: *** [Makefile:109: cleandocker] Error 1
```

7. What are the steps to reproduce the issue?
`sudo make cleandocker`

9. In what environment did you encounter the issue?
testing on terminal

11. Additional information you deem important:

### PR DESCRIPTION
**Pull request by SammyOina**
Signed-off-by: SammyOina <sammyoina@gmail.com>

### What does this do?
- New Feature: Introduced a new HTTP package that implements the `session.Handler` interface, enabling device connection, publishing, and authentication over HTTP.
- Refactor: Simplified the `MakeHandler` function in `http/api/transport.go` by removing unnecessary parameters and functionality.
- Refactor: Updated the `sendMessageEndpoint` function in `http/api/endpoint.go` to return `nil, nil`, simplifying its operation.
- Test: Revised tests in `http/api/endpoint_test.go` and `pkg/sdk/go/message_test.go` to accommodate changes in server setup and service creation.
- Chore: Renamed the package from "tracing" to "handler" in `pkg/messaging/handler/tracing.go` and updated the `NewHandler` function to `NewTracing`.
- Chore: Updated constants in `bootstrap/events/producer/streams_test.go` and `things/clients/service.go` for better code clarity and consistency.

### Which issue(s) does this PR fix/relate to?
Resolves https://github.com/ultravioletrs/issues/issues/267
Resolves https://github.com/mainflux/mainflux/issues/1942

### CODE REVIEW

File: cmd/http/main.go [hunk 1]
```diff
@@ -121,6 +126,19 @@ func main() {
 		return hs.Start()
 	})
 
+	switch {
+	case httpServerConfig.CertFile != "" || httpServerConfig.KeyFile != "":
+		logger.Info(fmt.Sprintf("%s service https server listening at %s:%s with TLS cert %s and key %s", svcName, httpServerConfig.Host, httpServerConfig.Port, httpServerConfig.CertFile, httpServerConfig.KeyFile))
+		if err := proxyHTTPS(ctx, httpServerConfig, logger, svc); err != nil {
+			logger.Error(fmt.Sprintf("failed to start proxy server with error: %v", err))
+		}
+	default:
```
Comment (rodneyosodo) [inline cmd/http/main.go]: Fix port mapping and URI
- This is the error from HTTP adapter
	```bash
	{"level":"error","message":"failed to start proxy server with error: parse \":81\": missing protocol scheme","ts":"2023-10-16T11:41:52.786991106Z"}
	```

- This is the error when trying to publish
	```bash
	unable to send messages to channels: HTTP failed to send message from thing 77f1cc97-2104-4ab0-be6e-5c88706b4196 to channel 0736b9f8-9293-49ec-bc92-4ff9550e33c3: Status: : Post "http://localhost:8008/channels/0736b9f8-9293-49ec-bc92-4ff9550e33c3/messages": read tcp [::1]:51840->[::1]:8008: read: connection reset by peer
	```

**Review by rodneyosodo — CHANGES_REQUESTED** [review]

**Comment by arvindh123** [comment]
@SammyOina 
I'm getting error in HTTP Adapter
```
2023-10-16 19:40:25 {"level":"info","message":"Successfully connected to things grpc server without TLS","ts":"2023-10-16T14:10:25.97790975Z"}
2023-10-16 19:40:25 {"level":"info","message":"http_adapter service http server listening at http-adapter:8008 without TLS","ts":"2023-10-16T14:10:25.994730439Z"}
2023-10-16 19:40:25 {"level":"info","message":"http_adapter service http server listening at http://localhost:81 without TLS","ts":"2023-10-16T14:10:25.9956256Z"}
2023-10-16 19:40:25 {"level":"info","message":"proxy HTTP shutdown at http://localhost:81","ts":"2023-10-16T14:10:25.995709404Z"}
2023-10-16 19:40:25 {"level":"error","message":"HTTP adapter service terminated: listen tcp: address http://localhost:81: too many colons in address","ts":"2023-10-16T14:10:25.996873955Z"}
2023-10-16 19:40:25 {"level":"error","message":"Error shutting down tracer provider: context canceled","ts":"2023-10-16T14:10:25.997285999Z"}
```

**Comment by arvindh123** [comment]
@SammyOina  @dborovcanin  @rodneyosodo 

I have some idea on this PR 

Please give your comments 

On looking the code, I understand, incoming messages are processed by proxy handler and there is no work for http service

So I have two proposal :

1.) Is possible to have tracing, logging, metering around the proxy handler and remove http service ? 

2.) In MQTT service,  mProxy does only authn, authz, or other kind of user defined work in incoming message and forward message to MQTT server. MQTT Server handles publishing, subscribing. 
Likewise can we do authn, authz and other checks in proxy handlers and let the service to do the publishing message in broker.

**Comment by SammyOina** [comment]
> @SammyOina  @dborovcanin  @rodneyosodo 
> 
> I have some idea on this PR 
> 
> Please give your comments 
> 
> On looking the code, I understand, incoming messages are processed by proxy handler and there is no work for http service
> 
> So I have two proposal :
> 
> 1.) Is possible to have tracing, logging, metering around the proxy handler and remove http service ? 
> 
> 2.) In MQTT service,  mProxy does only authn, authz, or other kind of user defined work in incoming message and forward message to MQTT server. MQTT Server handles publishing, subscribing. 
> Likewise can we do authn, authz and other checks in proxy handlers and let the service to do the publishing message in broker. 
> 


I had moved tracing, logging and metrics to PKG/messaging/handler

**Comment by arvindh123** [comment]
> > @SammyOina  @dborovcanin  @rodneyosodo
> > I have some idea on this PR
> > Please give your comments
> > On looking the code, I understand, incoming messages are processed by proxy handler and there is no work for http service
> > So I have two proposal :
> > 1.) Is possible to have tracing, logging, metering around the proxy handler and remove http service ?
> > 2.) In MQTT service,  mProxy does only authn, authz, or other kind of user defined work in incoming message and forward message to MQTT server. MQTT Server handles publishing, subscribing.
> > Likewise can we do authn, authz and other checks in proxy handlers and let the service to do the publishing message in broker.
> 
> I had moved tracing, logging and metrics to PKG/messaging/handler

Yes I saw the code.
Is possible to remove http server , mProxy proxies request to http sever , but the http server doesn't do nothing. 
So instead of forwarding the to http server which doesn't do nothing,  the request can be terminate with response code by the handler itself.

**Comment by SammyOina** [comment]
> > > @SammyOina  @dborovcanin  @rodneyosodo
> > > I have some idea on this PR
> > > Please give your comments
> > > On looking the code, I understand, incoming messages are processed by proxy handler and there is no work for http service
> > > So I have two proposal :
> > > 1.) Is possible to have tracing, logging, metering around the proxy handler and remove http service ?
> > > 2.) In MQTT service,  mProxy does only authn, authz, or other kind of user defined work in incoming message and forward message to MQTT server. MQTT Server handles publishing, subscribing.
> > > Likewise can we do authn, authz and other checks in proxy handlers and let the service to do the publishing message in broker.
> > 
> > 
> > I had moved tracing, logging and metrics to PKG/messaging/handler
> 
> Yes I saw the code. Is possible to remove http server , mProxy proxies request to http sever , but the http server doesn't do nothing. So instead of forwarding the to http server which doesn't do nothing, the request can be terminate with response code by the handler itself.

I think this is more of a question of what the mproxy library does, as it stands since it is a proxy it should allow to pass requests. Also, The current http server on mainflux doesn't do anything other than respond with HTTP status ok which is pretty much nothing. There is some remaining logic regarding content type because we can't put it on mproxy.

**Comment by dborovcanin** [comment]
> @SammyOina @dborovcanin @rodneyosodo
> 
> I have some idea on this PR
> 
> Please give your comments
> 
> On looking the code, I understand, incoming messages are processed by proxy handler and there is no work for http service
> 
> So I have two proposal :
> 
> 1.) Is possible to have tracing, logging, metering around the proxy handler and remove http service ?
> 
> 2.) In MQTT service, mProxy does only authn, authz, or other kind of user defined work in incoming message and forward message to MQTT server. MQTT Server handles publishing, subscribing. Likewise can we do authn, authz and other checks in proxy handlers and let the service to do the publishing message in broker.

1) It is, that's what @SammyOina did.
2) It does make sense, but it is not in the nature of the "proxy" concept here. It is the simplest possible HTTP server, though, so I would keep it as is for the time being.

File: cmd/http/main.go [hunk 2]
```diff
@@ -130,11 +147,60 @@ func main() {
...
+	logger.Info(fmt.Sprintf("%s service https server listening at %s:%s with TLS cert %s and key %s", svcName, cfg.Host, cfg.Port, cfg.CertFile, cfg.KeyFile))
+
+	select {
+	case <-ctx.Done():
+		logger.Info(fmt.Sprintf("proxy MQTT WS shutdown at %s", target))
+		return nil
+	case err := <-errCh:
+		return err
+	}
+}
```
Comment (dborovcanin) [inline cmd/http/main.go]: Use a single function with a param for TLS.

**Review by dborovcanin — CHANGES_REQUESTED** [review]

**Review by dborovcanin — APPROVED** [review]

**Review by drasko — APPROVED** [review]
LGTM

**Pull request merged**